src/SprykerEco/Zed/Unzer/Business/Payment/Filter/UnzerEnabledPaymentMethodFilter.php

```diff
@@ -0,0 +1,200 @@
+            }
+        }
+
+        return $activePaymentMethods;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\PaymentMethodsTransfer $paymentMethodsTransfer
+     *
+     * @return array<string|null>
+     */
+    protected function getPaymentMethodKeys(PaymentMethodsTransfer $paymentMethodsTransfer): array
+    {
+        return array_map(function (PaymentMethodTransfer $paymentMethodTransfer) {
+            return $paymentMethodTransfer->getPaymentMethodKey();
+        }, (array)$paymentMethodsTransfer->getMethods());
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\StoreTransfer $storeTransfer
+     * @param array<int> $unzerCredentialsTypes
+     *
+     * @throws \SprykerEco\Zed\Unzer\Business\Exception\UnzerException
+     *
+     * @return \Generated\Shared\Transfer\UnzerKeypairTransfer
+     */
+    protected function getUnzerKeypair(StoreTransfer $storeTransfer, array $unzerCredentialsTypes): UnzerKeypairTransfer
+    {
+        $unzerCredentialsCriteriaTransfer = (new UnzerCredentialsCriteriaTransfer())->setUnzerCredentialsConditions(
+            (new UnzerCredentialsConditionsTransfer())
+                ->addStoreName($storeTransfer->getNameOrFail())
+                ->setTypes($unzerCredentialsTypes),
+        );
+
+        $unzerCredentialsTransfer = $this->unzerReader->findUnzerCredentialsByCriteria($unzerCredentialsCriteriaTransfer);
+        if ($unzerCredentialsTransfer === null) {
+            throw new UnzerException('Unzer Credentials for current Store configuration not found.');
+        }
+
+        return $unzerCredentialsTransfer->getUnzerKeypairOrFail();
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Generated\Shared\Transfer\UnzerCredentialsCollectionTransfer
+     */
+    protected function getMerchantUnzerCredentialsCollection(QuoteTransfer $quoteTransfer): UnzerCredentialsCollectionTransfer
+    {
+        $merchantReferences = $this->getUniqueMerchantReferences($quoteTransfer);
+        $unzerCredentialsCriteriaTransfer = (new UnzerCredentialsCriteriaTransfer())->setUnzerCredentialsConditions(
+            (new UnzerCredentialsConditionsTransfer())
+                ->setMerchantReferences($merchantReferences)
+                ->addStoreName($quoteTransfer->getStoreOrFail()->getNameOrFail()),
+        );
+
+        return $this->unzerReader->getUnzerCredentialsCollectionByCriteria($unzerCredentialsCriteriaTransfer);
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return array<array-key, string>
+     */
+    protected function getUniqueMerchantReferences(QuoteTransfer $quoteTransfer): array
+    {
+        $merchantReferences = [];
+
+        foreach ($quoteTransfer->getItems() as $itemTransfer) {
+            if ($itemTransfer->getMerchantReference() !== null) {
+                $merchantReferences[] = $itemTransfer->getMerchantReferenceOrFail();
```

Comment:
```suggestion
                $merchantReferences[$itemTransfer->getMerchantReferenceOrFail()] = $itemTransfer->getMerchantReferenceOrFail();
```

in this case `array_unique` won't be needed